File: DS/C#/DoublyLinkedList/DoublyLinkedList.cs
//using System.Collections;

public class DoublyLinkedListNode<T>
{
	public T Value { get; set; }
	public DoublyLinkedListNode<T> Next { get; set; }
	public DoublyLinkedListNode<T> Previous { get; set; }
	
	public DoublyLinkedListNode(T value) 
	{
		Value = value;
	}
}

public class DoublyLinkedList<T> : System.Collections.Generic.ICollection<T>
{ 
	public DoublyLinkedListNode<T> Head { get; private set; }
	public DoublyLinkedListNode<T> Tail { get; private set; }
	public int Count { get; private set; }
	
	public void AddHead(DoublyLinkedListNode<T> node) 
	{
		DoublyLinkedListNode<T> temp = Head;
		
		Head = node;
		Head.Next = temp;
		
		if (Count == 0) {
			Tail = Head;
		} else {
			temp.Previous = Head;
		}
		
		Count++;
	}
	
	public void AddTail(DoublyLinkedListNode<T> node) 
	{
		if (Count == 0) {
			Head = Tail;
		} else {
			Head.Next = node;
			Node.Previous = Tail;
		}
		
		Tail = node;
		Count++;
	}
}

File: DS/Java/DoublyLinkedList/DoublyLinkedList.java

public class DoublyLinkedList {
	Node head;
	Node tail;
	int count;
	
	class Node {
		int value;
		Node previous;
		Node next;
		
		Node(int data) {
			this.value = data;
		}
		
	}
	
	// creation operation
	public void addFirst(LinkedList<T> node) {
		temp = head;
		head = node;
		head.next = temp;

		count++

		if (count == 1) {
			tail = head;
		} else {
			temp.previous = head;
		}
	}

	public void addLast(LinkedList<T> node) {
		if (count == 0) {
			head = node;
		} else {
			tail.next = node;
			node.previous = tail;
		}

		tail = node;

		count++
	}


	// remove operation
	public void removeFirst() {
		if (count != 0) {
			head = head.next;
			count--
			if (count == 0) {
				tail = null;
			} else {
				head.previous = null;
			}

		}

	}

	public void removeLast() {
		if (count != 0) {
			if (count === 1) {
				head = null;
				tail = null;
			} else {
				tail.previous.next = null;
				tail = tail.previous;
			}
			count--
		}

	}


	public boolean remove(T item) {
		prev = null;
		current = head;

		while (current != null) {
			if (current.value.equals(item)) {
				if (prev !== null) {
					prev.next = current.next;
					if (current.next == null) {
						tail = prev;
					} else {
						current.next.previous = prev;
					}
					count--;
				} else {
					removeFirst();
				}
				return true;
			}
			prev = current;
			current = current.next;
		}
		return false;
	}


	// Enumerate
	public void enumerate() {
		current = head;
		while (current != null) {
			System.out.println(current.value);
			current = current.next;
		}

	}

	// Contains
	public boolean contains(T item) {
		current = head;
		while (current != null) {
			if (current.value.equals(item)) {
				return true;
			}
			current = current.next;
		}
		return false;
	}

	// Copy
	public void copyTo(T[] array, int index) {
		current = head;
		while (current != null) {
			array[index++] = current.value;
			current = current.next;
		}
	}

	public void clear() {
		head = null;
		tail = null;
		count = 0;
	}

}
